src/Compilers/CSharp/Portable/FlowAnalysis/VariableUsePass.cs

```diff
@@ -0,0 +1,111 @@
+        private readonly SourceAssemblySymbol _sourceAssembly;
+
+        /// <summary>
+        /// Create a field-only visitor.
+        /// </summary>
+        /// <remarks>
+        /// Even though this class looks stateless, it cannot be shared/cached on the CSharpCompilation.
+        /// Specifically, BoundTreeWalkerWithStackGuard has an int _recursionDepth that breaks in multithreaded scenarios.
+        /// </remarks>
+        public FieldUsePass(SourceAssemblySymbol sourceAssembly)
+        {
+            _sourceAssembly = sourceAssembly;
+        }
+
+        /// <summary>
+        /// Diagnostics are used when a CancelledByStackGuardException is thrown.
+        /// </summary>
+        public void Analyze(BoundNode node, DiagnosticBag diagnostics)
+        {
+            try
+            {
+                Visit(node);
+            }
+            catch (CancelledByStackGuardException ex)
+            {
+                ex.AddAnError(diagnostics);
+            }
+        }
+
+        public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+        {
+            _sourceAssembly.NoteFieldAccess(node.FieldSymbol, true, false);
+            return base.VisitFieldAccess(node);
+        }
+    }
+
+    internal sealed class LocalVariableUsePass : BoundTreeWalkerWithStackGuard
+    {
+        /// <summary>
+        /// Variables that were used anywhere, in the sense required to suppress warnings about
+        /// unused variables.
+        /// </summary>
+        private readonly HashSet<LocalSymbol> _usedVariables;
+
+        /// <summary>
+        /// Local functions that were used anywhere, in the sense required to suppress warnings
+        /// about unused local functions.
+        /// </summary>
+        private readonly HashSet<LocalFunctionSymbol> _usedLocalFunctions;
+
+        /// <summary>
+        /// Create a visitor for use within a method (e.g. local functions).
+        /// Keep track of what local variables are referenced, by placing them in the HashSets.
+        /// Can throw <see cref="BoundTreeVisitor.CancelledByStackGuardException"/>.
+        /// </summary>
+        public LocalVariableUsePass(HashSet<LocalSymbol> usedVariables, HashSet<LocalFunctionSymbol> usedLocalFunctions)
+        {
+            _usedVariables = usedVariables;
+            _usedLocalFunctions = usedLocalFunctions;
+        }
+
+        public override BoundNode VisitLocal(BoundLocal node)
+        {
+            _usedVariables.Add(node.LocalSymbol);
+            return base.VisitLocal(node);
+        }
+
+        private void VisitMethodReference(MethodSymbol symbol)
+        {
+            if (symbol?.MethodKind == MethodKind.LocalFunction)
+            {
+                // Make sure we use the unconstructed local function, if it is generic
+                _usedLocalFunctions.Add((LocalFunctionSymbol)symbol.OriginalDefinition);
+            }
+        }
+
+        public override BoundNode VisitCall(BoundCall node)
```

Comment:
I don't think so. Look at `InvalidLocfuncReferencesMarkUsed`, there's examples like: `void DelCreation(Func<string> s = new Func<string>(Func3)) { }`, which pass without visiting delegate creations. From my understanding, `DelegateCreationExpression`s include a MethodGroup as a child, which is visited.